File: packages/veil-sdk-js/src/app_plugins.ts
import { decodeAppEnvelope, extractReferences } from "./app_schemas";
import type { AppEnvelope } from "./app_schemas";
import type { VeilClient, VeilClientPlugin } from "./client";

export type RootTagResolver = (
  objectRootHex: string,
  envelope?: AppEnvelope,
) => string | null;

export interface AutoFetchPluginOptions {
  resolveTagForRoot?: RootTagResolver;
  priority?: number;
}

export function createAutoFetchPlugin(
  options: AutoFetchPluginOptions = {},
): VeilClientPlugin {
  const priority = options.priority ?? 1;
  return {
    onObject(client: VeilClient, objectRootHex: string, objectBytes: Uint8Array) {
      let envelope: AppEnvelope;
      try {
        envelope = decodeAppEnvelope(objectBytes);
      } catch {
        return;
      }
      const refs = extractReferences(envelope);
      for (const tagHex of refs.chunkTagHexes) {
        client.subscribe(tagHex);
      }

      const resolveTag = options.resolveTagForRoot;
      const allRoots = [
        ...refs.parentRoots,
        ...refs.threadRoots,
        ...refs.chunkRoots,
      ];
      for (const root of allRoots) {
        if (priority > 0) {
          client.prioritizeObjectRoot(root, priority);
        }
        if (resolveTag) {
          const tag = resolveTag(root, envelope);
          if (tag) {
            client.subscribe(tag);
          }
        }
      }
    },
  };
}

export interface ThreadContextPluginOptions {
  resolveTagForRoot: RootTagResolver;
  priority?: number;
}

export function createThreadContextPlugin(
  options: ThreadContextPluginOptions,
): VeilClientPlugin {
  const priority = options.priority ?? 1;
  return {
    onObject(client: VeilClient, objectRootHex: string, objectBytes: Uint8Array) {
      let envelope: AppEnvelope;
      try {
        envelope = decodeAppEnvelope(objectBytes);
      } catch {
        return;
      }
      const refs = extractReferences(envelope);
      const roots = [...refs.parentRoots, ...refs.threadRoots];
      for (const root of roots) {
        if (priority > 0) {
          client.prioritizeObjectRoot(root, priority);
        }
        const tag = options.resolveTagForRoot(root, envelope);
        if (tag) {
          client.subscribe(tag);
        }
      }
    },
  };
}

File: packages/veil-sdk-dart/lib/src/plugins/app_plugins.dart
import "dart:typed_data";

import "../client.dart";
import "../models/app_schemas.dart";

typedef RootTagResolver =
    String? Function(String objectRootHex, AppEnvelope? envelope);

class AutoFetchPlugin implements VeilClientPlugin {
  final RootTagResolver? resolveTagForRoot;
  final int priority;

  const AutoFetchPlugin({this.resolveTagForRoot, this.priority = 1});

  @override
  void onObject(
    VeilClient client,
    String objectRootHex,
    Uint8List objectBytes,
  ) {
    AppEnvelope? envelope;
    try {
      envelope = decodeAppEnvelope(objectBytes);
    } catch (_) {
      return;
    }
    final refs = extractReferences(envelope);
    for (final tag in refs.chunkTagHexes) {
      client.subscribe(tag);
    }

    final allRoots = <String>[
      ...refs.parentRoots,
      ...refs.threadRoots,
      ...refs.chunkRoots,
    ];
    for (final root in allRoots) {
      if (priority > 0) {
        client.prioritizeObjectRoot(root, priority: priority);
      }
      if (resolveTagForRoot != null) {
        final tag = resolveTagForRoot!(root, envelope);
        if (tag != null) {
          client.subscribe(tag);
        }
      }
    }
  }
}

class ThreadContextPlugin implements VeilClientPlugin {
  final RootTagResolver resolveTagForRoot;
  final int priority;

  const ThreadContextPlugin({
    required this.resolveTagForRoot,
    this.priority = 1,
  });

  @override
  void onObject(
    VeilClient client,
    String objectRootHex,
    Uint8List objectBytes,
  ) {
    AppEnvelope? envelope;
    try {
      envelope = decodeAppEnvelope(objectBytes);
    } catch (_) {
      return;
    }
    final refs = extractReferences(envelope);
    final roots = <String>[...refs.parentRoots, ...refs.threadRoots];
    for (final root in roots) {
      if (priority > 0) {
        client.prioritizeObjectRoot(root, priority: priority);
      }
      final tag = resolveTagForRoot(root, envelope);
      if (tag != null) {
        client.subscribe(tag);
      }
    }
  }
}
